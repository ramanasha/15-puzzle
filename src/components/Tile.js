import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import classnames from 'classnames';

class Tile extends Component {
  static propTypes = {
    tileId: PropTypes.number.isRequired,
    width: PropTypes.number.isRequired,
    height: PropTypes.number.isRequired,
    left: PropTypes.number.isRequired,
    top: PropTypes.number.isRequired,
    number: PropTypes.number,
    onClick: PropTypes.func,
    correct: PropTypes.bool,
    visible: PropTypes.bool,
  };

  static defaultProps = {
    number: 0,
    correct: false,
    visible: true,
  };

  render() {
    const { className, number, visible } = this.props;
    const classes = classnames({
      [className]: true,
      'tile-invisible': !visible,
    });

    return (
      <div className={classes} onClick={this.onClick}>
        <span className="tile-number">{number}</span>
      </div>
    );
  }

  onClick = () => {
    this.props.onClick({
      number: this.props.number,
    });
  };
}

export default styled(Tile)`
  border: 1px solid #FFD1AA;
  position: absolute;
  width: ${props => props.width}px;
  height: ${props => props.height}px;
  left: ${props => props.left}px;
  top: ${props => props.top}px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  cursor: pointer;
  background-color: ${props => props.correct ? '#226666' : '#D4726A'};
  transition-property: top, left, background-color;
  transition-duration: .300s;
  transition-timing-function: ease-in;

  .tile-number {
    color: #FFD1AA;
    font-weight: 400;
    fontSize: 1.8em;
    user-select: none;
  }

  &.tile-invisible {
    display: none;
  }
`;

// Export this for easier unit-testing
export { Tile as Component };

export const propTypes = Tile.propTypes;
