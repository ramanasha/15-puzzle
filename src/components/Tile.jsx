import React, { Component } from 'react';
import styled from 'styled-components';
import classnames from 'classnames';
import PropTypes from 'prop-types';

class Tile extends Component {
  onClick = () => {
    this.props.onClick({
      number: this.props.number
    });
  };

  render() {
    const { className, number, visible } = this.props;
    const classes = classnames({
      [className]: true,
      'tile-invisible': !visible
    });

    return (
      <div className={classes} onClick={this.onClick}>
        <span className="tile-number">{number}</span>
      </div>
    );
  }
}

Tile.propTypes = {
  tileId: PropTypes.number.isRequired,
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
  left: PropTypes.number.isRequired,
  top: PropTypes.number.isRequired,
  number: PropTypes.number,
  onClick: PropTypes.func.isRequired,
  correct: PropTypes.bool,
  visible: PropTypes.bool
};

Tile.defaultProps = {
  number: 0,
  correct: false,
  visible: true
};

export default styled(Tile)`
  border: 1px solid #ffd1aa;
  position: absolute;
  width: ${props => props.width}px;
  height: ${props => props.height}px;
  left: ${props => props.left}px;
  top: ${props => props.top}px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  cursor: pointer;
  background-color: ${props => (props.correct ? '#226666' : '#D4726A')};
  transition-property: top, left, background-color;
  transition-duration: 0.3s;
  transition-timing-function: ease-in;

  .tile-number {
    color: #ffd1aa;
    font-weight: 400;
    font-size: 1.8em;
    user-select: none;
  }

  &.tile-invisible {
    display: none;
  }
`;

// Export this for easier unit-testing
export { Tile as Component };

export const propTypes = Tile.propTypes;
