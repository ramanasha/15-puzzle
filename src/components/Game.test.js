import renderer from 'react-test-renderer';
import React from 'react';
import { shallow, mount } from 'enzyme';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import injectTapEventPlugin from 'react-tap-event-plugin';
import Game from './Game';
import Tile from './Tile';
import * as GameStates from '../lib/game-status';

injectTapEventPlugin();

describe('Game', () => {
  const NUMBERS = [1, 4, 6, 7, 2, 8, 14, 13, 10, 9, 3, 5, 12, 11];

  describe('render', () => {
    it('should match a shallow render snapshot', () => {
      const wrapper = shallow(
        <Game
          numbers={NUMBERS}
          className="random-cls"
          onResetClick={() => {}}
          onTileClick={() => {}}
          original={NUMBERS}
        />,
      );
      expect(wrapper.getNode()).toMatchSnapshot();
    });

    it('should ');
  });

  describe('callbacks', () => {
    it('should restart the game every time new numbers are passed', () => {
      const wrapper = shallow(
        <Game
          numbers={NUMBERS}
          className="random-cls"
          onResetClick={() => {}}
          onTileClick={() => {}}
          original={NUMBERS}
        />,
      );

      // Start the game
      wrapper.setState({
        gameState: GameStates.GAME_STARTED,
      });
      const oldTiles = wrapper.state('tiles');

      // Change numbers, this triggers new tiles
      wrapper.setProps({
        numbers: [2, 5, 1, 3, 4, 8, 14, 13, 10, 9, 3, 5, 12, 11],
      });

      expect(wrapper.state('tiles')).not.toEqual(oldTiles);
      expect(wrapper.state).toMatchObject({
        gameState: GameStates.GAME_IDLE,
        moves: 0,
        seconds: 0,
      });
    });

    it('should not respond to clicks if game is over', () => {
      const wrapper = mount(
        <MuiThemeProvider>
          <Game
            numbers={NUMBERS}
            className="random-cls"
            onResetClick={() => {}}
            onTileClick={() => {}}
            original={NUMBERS}
          />
        </MuiThemeProvider>,
      );

      wrapper.setState({
        gameState: GameStates.GAME_OVER,
      });

      const expectedTiles = wrapper.state('tiles');

      // Click the tile that can move, check state that it didn't
      wrapper.find({ number: 11 }).simulate('click');

      expect(wrapper.state('tiles')).toEqual(expectedTiles);
    });
  });
});
