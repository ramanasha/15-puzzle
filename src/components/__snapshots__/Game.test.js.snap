exports[`Game render should match a shallow render snapshot 1`] = `
<div
  className="random-cls">
  <Menu
    moves={0}
    onResetClick={[Function]}
    seconds={0} />
  <Styled(Grid)
    gridSize={4}
    onTileClick={[Function]}
    tileSize={90}
    tiles={
      Array [
        Object {
          "column": 0,
          "height": 90,
          "left": 0,
          "number": 1,
          "row": 0,
          "tileId": 0,
          "top": 0,
          "width": 90,
        },
        Object {
          "column": 1,
          "height": 90,
          "left": 90,
          "number": 4,
          "row": 0,
          "tileId": 1,
          "top": 0,
          "width": 90,
        },
        Object {
          "column": 2,
          "height": 90,
          "left": 180,
          "number": 6,
          "row": 0,
          "tileId": 2,
          "top": 0,
          "width": 90,
        },
        Object {
          "column": 3,
          "height": 90,
          "left": 270,
          "number": 7,
          "row": 0,
          "tileId": 3,
          "top": 0,
          "width": 90,
        },
        Object {
          "column": 0,
          "height": 90,
          "left": 0,
          "number": 2,
          "row": 1,
          "tileId": 4,
          "top": 90,
          "width": 90,
        },
        Object {
          "column": 1,
          "height": 90,
          "left": 90,
          "number": 8,
          "row": 1,
          "tileId": 5,
          "top": 90,
          "width": 90,
        },
        Object {
          "column": 2,
          "height": 90,
          "left": 180,
          "number": 14,
          "row": 1,
          "tileId": 6,
          "top": 90,
          "width": 90,
        },
        Object {
          "column": 3,
          "height": 90,
          "left": 270,
          "number": 13,
          "row": 1,
          "tileId": 7,
          "top": 90,
          "width": 90,
        },
        Object {
          "column": 0,
          "height": 90,
          "left": 0,
          "number": 10,
          "row": 2,
          "tileId": 8,
          "top": 180,
          "width": 90,
        },
        Object {
          "column": 1,
          "height": 90,
          "left": 90,
          "number": 9,
          "row": 2,
          "tileId": 9,
          "top": 180,
          "width": 90,
        },
        Object {
          "column": 2,
          "height": 90,
          "left": 180,
          "number": 3,
          "row": 2,
          "tileId": 10,
          "top": 180,
          "width": 90,
        },
        Object {
          "column": 3,
          "height": 90,
          "left": 270,
          "number": 5,
          "row": 2,
          "tileId": 11,
          "top": 180,
          "width": 90,
        },
        Object {
          "column": 0,
          "height": 90,
          "left": 0,
          "number": 12,
          "row": 3,
          "tileId": 12,
          "top": 270,
          "width": 90,
        },
        Object {
          "column": 1,
          "height": 90,
          "left": 90,
          "number": 11,
          "row": 3,
          "tileId": 13,
          "top": 270,
          "width": 90,
        },
      ]
    } />
  <Dialog
    actions={
      Array [
        <FlatButton
          disabled={false}
          fullWidth={false}
          label="Close"
          labelPosition="after"
          labelStyle={Object {}}
          onKeyboardFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onTouchStart={[Function]}
          onTouchTap={[Function]}
          primary={false}
          secondary={false} />,
      ]
    }
    autoDetectWindowHeight={true}
    autoScrollBodyContent={false}
    modal={false}
    onRequestClose={[Function]}
    open={false}
    repositionOnUpdate={true}
    title="Congrats!">
    You\'ve solved the puzzle in
     
    0
     
    moves in
     
    0
     
    seconds!
  </Dialog>
</div>
`;

exports[`Game render should render correctly given a default grid 1`] = `
<div
  className="random-cls">
  <Menu
    moves={0}
    onResetClick={[Function]}
    seconds={0} />
  <Styled(Grid)
    gridSize={4}
    onTileClick={[Function]}
    tileSize={90}
    tiles={
      Array [
        Object {
          "column": 0,
          "height": 90,
          "left": 0,
          "number": 1,
          "row": 0,
          "tileId": 0,
          "top": 0,
          "width": 90,
        },
        Object {
          "column": 1,
          "height": 90,
          "left": 90,
          "number": 4,
          "row": 0,
          "tileId": 1,
          "top": 0,
          "width": 90,
        },
        Object {
          "column": 2,
          "height": 90,
          "left": 180,
          "number": 6,
          "row": 0,
          "tileId": 2,
          "top": 0,
          "width": 90,
        },
        Object {
          "column": 3,
          "height": 90,
          "left": 270,
          "number": 7,
          "row": 0,
          "tileId": 3,
          "top": 0,
          "width": 90,
        },
        Object {
          "column": 0,
          "height": 90,
          "left": 0,
          "number": 2,
          "row": 1,
          "tileId": 4,
          "top": 90,
          "width": 90,
        },
        Object {
          "column": 1,
          "height": 90,
          "left": 90,
          "number": 8,
          "row": 1,
          "tileId": 5,
          "top": 90,
          "width": 90,
        },
        Object {
          "column": 2,
          "height": 90,
          "left": 180,
          "number": 14,
          "row": 1,
          "tileId": 6,
          "top": 90,
          "width": 90,
        },
        Object {
          "column": 3,
          "height": 90,
          "left": 270,
          "number": 13,
          "row": 1,
          "tileId": 7,
          "top": 90,
          "width": 90,
        },
        Object {
          "column": 0,
          "height": 90,
          "left": 0,
          "number": 10,
          "row": 2,
          "tileId": 8,
          "top": 180,
          "width": 90,
        },
        Object {
          "column": 1,
          "height": 90,
          "left": 90,
          "number": 9,
          "row": 2,
          "tileId": 9,
          "top": 180,
          "width": 90,
        },
        Object {
          "column": 2,
          "height": 90,
          "left": 180,
          "number": 3,
          "row": 2,
          "tileId": 10,
          "top": 180,
          "width": 90,
        },
        Object {
          "column": 3,
          "height": 90,
          "left": 270,
          "number": 5,
          "row": 2,
          "tileId": 11,
          "top": 180,
          "width": 90,
        },
        Object {
          "column": 0,
          "height": 90,
          "left": 0,
          "number": 12,
          "row": 3,
          "tileId": 12,
          "top": 270,
          "width": 90,
        },
        Object {
          "column": 1,
          "height": 90,
          "left": 90,
          "number": 11,
          "row": 3,
          "tileId": 13,
          "top": 270,
          "width": 90,
        },
      ]
    } />
  <Dialog
    actions={
      Array [
        <FlatButton
          disabled={false}
          fullWidth={false}
          label="Close"
          labelPosition="after"
          labelStyle={Object {}}
          onKeyboardFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onTouchStart={[Function]}
          onTouchTap={[Function]}
          primary={false}
          secondary={false} />,
      ]
    }
    autoDetectWindowHeight={true}
    autoScrollBodyContent={false}
    modal={false}
    onRequestClose={[Function]}
    open={false}
    repositionOnUpdate={true}
    title="Congrats!">
    You've solved the puzzle in
     
    0
     
    moves in
     
    0
     
    seconds!
  </Dialog>
</div>
`;
